### ISSUE TITLE
Can't iterate over #vars without HttpSession

### ISSUE THREAD

**Issue by hho** [thymeleaf/thymeleaf#349]
I'm just getting started with Spring Boot and Thymeleaf and tried to dump all the `#vars` just on my index page.

``` xml
<table class="table table-bordered table-hover">
    <tr th:each="c : ${#vars}">
        <td th:text="${c.key}"></td>
        <td th:text="${c.value}"></td>
    </tr>
</table>
```

This fails with a NullPointerException here…
https://github.com/thymeleaf/thymeleaf/blob/thymeleaf-2.1.4.RELEASE/src/main/java/org/thymeleaf/context/WebSessionVariablesMap.java#L276
…because I don't seem to have a `HttpSession` yet.

Versions:

```
thymeleaf         2.1.4.RELEASE
thymeleaf-spring4 2.1.4.RELEASE
spring-boot-*     1.2.1.RELEASE
JDK 1.8.0_31-b13
Windows 7 [6.1.7601]
```

Stacktrace:

```
java.lang.NullPointerException: null
    at org.thymeleaf.context.WebSessionVariablesMap.hashCode(WebSessionVariablesMap.java:276)
    at java.util.Objects.hashCode(Objects.java:98)
    at java.util.HashMap$Node.hashCode(HashMap.java:296)
    at java.util.HashMap.hash(HashMap.java:338)
    at java.util.HashMap.put(HashMap.java:611)
    at java.util.HashSet.add(HashSet.java:219)
    at java.util.AbstractCollection.addAll(AbstractCollection.java:344)
    at java.util.HashSet.<init>(HashSet.java:119)
    at org.thymeleaf.context.WrappingVariablesMap.entrySet(WrappingVariablesMap.java:219)
    at org.thymeleaf.util.EvaluationUtil.evaluateAsIterable(EvaluationUtil.java:128)
    at org.thymeleaf.processor.attr.AbstractIterationAttrProcessor.processAttribute(AbstractIterationAttrProcessor.java:109)
    at org.thymeleaf.processor.attr.AbstractAttrProcessor.doProcess(AbstractAttrProcessor.java:87)
    at org.thymeleaf.processor.AbstractProcessor.process(AbstractProcessor.java:212)
    at org.thymeleaf.dom.Node.applyNextProcessor(Node.java:1017)
    at org.thymeleaf.dom.Node.processNode(Node.java:972)
    at org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:695)
    at org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:668)
    at org.thymeleaf.dom.Node.processNode(Node.java:990)
    at org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:695)
    at org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:668)
    at org.thymeleaf.dom.Node.processNode(Node.java:990)
    at org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:695)
    at org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:668)
    at org.thymeleaf.dom.Node.processNode(Node.java:990)
    at org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:695)
    at org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:668)
    at org.thymeleaf.dom.Node.processNode(Node.java:990)
    at org.thymeleaf.dom.NestableNode.computeNextChild(NestableNode.java:695)
    at org.thymeleaf.dom.NestableNode.doAdditionalProcess(NestableNode.java:668)
    at org.thymeleaf.dom.Node.processNode(Node.java:990)
    at org.thymeleaf.dom.Document.process(Document.java:93)
    at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1155)
    at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1060)
    at org.thymeleaf.TemplateEngine.process(TemplateEngine.java:1011)
    at org.thymeleaf.spring4.view.ThymeleafView.renderFragment(ThymeleafView.java:335)
    at org.thymeleaf.spring4.view.ThymeleafView.render(ThymeleafView.java:190)
    at org.springframework.web.servlet.DispatcherServlet.render(DispatcherServlet.java:1228)
    at org.springframework.web.servlet.DispatcherServlet.processDispatchResult(DispatcherServlet.java:1011)
    at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:955)
    at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:877)
    at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:966)
    at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:857)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:618)
    at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:842)
    at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:291)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at org.springframework.web.filter.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:77)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at org.springframework.web.filter.CharacterEncodingFilter.doFilterInternal(CharacterEncodingFilter.java:88)
    at org.springframework.web.filter.OncePerRequestFilter.doFilter(OncePerRequestFilter.java:107)
    at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:239)
    at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:206)
    at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:219)
    at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:106)
    at org.apache.catalina.authenticator.AuthenticatorBase.invoke(AuthenticatorBase.java:501)
    at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:142)
    at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:79)
    at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:88)
    at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:537)
    at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1085)
    at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:658)
    at org.apache.coyote.http11.Http11NioProtocol$Http11ConnectionHandler.process(Http11NioProtocol.java:222)
    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1556)
    at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1513)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
    at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by mreiterer** [comment]
same here. seems to be a problem in 2.1.4. Trying the same in 2.1.3 works

**Comment by danielfernandez** [comment]
Fixed in 2.1. Does not affect the 3.0 branch.

Already in `2.1.5-SNAPSHOT`

**Comment by aaaaarrrgghhh** [comment]
I'm having this same error with ThymeLeaf 2.1.4 in Spring Boot, with Eclipse, Tomcat 8.
Any workarounds please?

**Comment by ultraq** [comment]
Given the comment above, I'd say the fix is to try the 2.1.5-SNAPSHOT version and see if that works.

**Comment by aaaaarrrgghhh** [comment]
But 2.1.4 is the latest release available on the website!

**Comment by ultraq** [comment]
True, but if you configure Maven to download from the snapshots repository, you can give 2.1.5-SNAPSHOT a try too.  We have a guide for that here: http://www.thymeleaf.org/faq.html#use-snapshot

If you're not using Maven or a Maven-compatible dependency management tool, then you can download it from the Maven snapshot repository here: https://oss.sonatype.org/content/repositories/snapshots/org/thymeleaf/thymeleaf/2.1.5-SNAPSHOT/  You'll be after the latest 2.1.5-SNAPSHOT jar file, probably the one named `thymeleaf-2.1.5-20160102.163400-6.jar` as of writing.

**Comment by aaaaarrrgghhh** [comment]
Ok...any ideas if its been tested with Spring Boot..

**Comment by ultraq** [comment]
Not sure, but given that this is a patch change over 2.1.4, then the interfaces shouldn't have changed and so I'd expect it to continue to work with Spring Boot.  So you'll really have to just try it yourself.  The tests for the latest 2.1 development code are also green if that's any comfort.

**Comment by aaaaarrrgghhh** [comment]
So this is weird. I changed to version 2.1.5-SNAPSHOT. Same error
Exception evaluating SpringEL expression: "question.questionText" (ehcForm:14).
BUT, when i change the template variable to question.ID, it still shows me the same error:
Exception evaluating SpringEL expression: "question.questionText" (ehcForm:14).

I've tried disabling caches, restarting Tomcat, used Safari, Chrome - same result! There is no "questionText" now at that Line 14, and I'm still getting the same error! I've verified that "question" is not null in my Controller before putting it into the Model. 
I understand this isn't the forum, but just thought I'd put it out here...
